x-pack/plugins/apm/server/routes/transactions/trace_samples/index.ts

```diff
@@ -36,20 +49,62 @@ export async function getTransactionTraceSamples({
   start: number;
   end: number;
 }) {
-  return withApmSpan('get_transaction_trace_samples', async () => {
-    return await getTraceSamples({
-      environment,
-      kuery,
-      serviceName,
-      transactionName,
-      transactionType,
-      transactionId,
-      traceId,
-      sampleRangeFrom,
-      sampleRangeTo,
-      setup,
-      start,
-      end,
+  return withApmSpan('get_trace_samples', async () => {
+    const { apmEventClient } = setup;
+
+    const commonFilters = [
+      { term: { [SERVICE_NAME]: serviceName } },
+      { term: { [TRANSACTION_TYPE]: transactionType } },
+      { term: { [TRANSACTION_NAME]: transactionName } },
+      ...rangeQuery(start, end),
+      ...environmentQuery(environment),
+      ...kqlQuery(kuery),
+    ] as QueryDslQueryContainer[];
+
+    if (sampleRangeFrom !== undefined && sampleRangeTo !== undefined) {
+      commonFilters.push({
+        range: {
+          'transaction.duration.us': {
+            gte: sampleRangeFrom,
+            lte: sampleRangeTo,
+          },
+        },
+      });
+    }
+
+    const response = await apmEventClient.search('get_trace_samples_hits', {
+      apm: {
+        events: [ProcessorEvent.transaction],
+      },
+      _source: [TRANSACTION_ID, TRACE_ID, '@timestamp'],
+      body: {
+        query: {
+          bool: {
+            filter: [
+              ...commonFilters,
+              { term: { [TRANSACTION_SAMPLED]: true } },
+            ],
+            should: [
+              { term: { [TRACE_ID]: traceId } },
+              { term: { [TRANSACTION_ID]: transactionId } },
+            ] as QueryDslQueryContainer[],
+          },
+        },
+        size: TRACE_SAMPLES_SIZE,
+      },
     });
+
+    const traceSamples = orderBy(
+      response.hits.hits.map((hit) => ({
+        score: hit._score,
+        timestamp: hit._source['@timestamp'],
+        transactionId: hit._source.transaction.id,
+        traceId: hit._source.trace.id,
+      })),
+      ['score', 'timestamp'],
+      ['desc', 'desc']
+    );
```

Comment:
Sorting by score (like we already do) and then timestamp

Reply:
Why are we sorting on the score? Aren't we just filtering the transactions and don't do searches that may impact the score depending on how well the transaction matches the query. I might be missing something 🙂 

Also, it seems we're sorting in Kibana rather than specifying the sort order on the ES query which might be more efficient.

Reply:
> Why are we sorting on the score?

The query has a `should` clause which can match exactly one document. We want this document to be the first (regardless of timestamp). The remaining documents will all have a score of 0, and should be sorted by timestamp.

> Also, it seems we're sorting in Kibana rather than specifying the sort order on the ES query which might be more efficient.

My intuition is telling me that it is faster to sort 500 documents on the Kibana server than sorting potentially millions of documents in ES. I didn't benchmark this so I could be wrong.
Regardless, I'm not sure we want to sort in ES, since this would lead to strictly only showing 500 most recent samples. I even considered adding a sort on `trace.id` in ES to get a somewhat random distribution over the selected time range while still sorting by time on the server to ensure they are in order when displaying to the user.